### PR TITLE
I18n: Add locale to user preference

### PR DESCRIPTION
**Pull request by joshhunt**
**What this PR does / why we need it**:

Starts to add locale preference. Future PRs will add this to Org preferences and also the custom.ini config.

Because this is optional in the UI, we're not adding this particular part behind a feature flag, despite the UI for this being feature flagged (when its added)

**Which issue(s) this PR fixes**:

Part of #49424

### CODE REVIEW

File: pkg/services/preference/prefimpl/pref.go [hunk 1]
```diff
@@ -106,9 +110,14 @@ func (s *Service) Save(ctx context.Context, cmd *pref.SavePreferenceCommand) err
 	preference.Theme = cmd.Theme
 	preference.Updated = time.Now()
 	preference.Version += 1
-	preference.JSONData = &pref.PreferenceJSONData{}
 	preference.HomeDashboardID = cmd.HomeDashboardID
 
+	if preference.JSONData == nil {
+		preference.JSONData = &pref.PreferenceJSONData{}
+	}
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply pkg/services/preference/prefimpl/pref.go:117]: Actually, thinking about it before, I'm thinking maybe that behavior was intentional?

[the remark this answers is not in the document]
Comment (joshhunt) [reply pkg/services/preference/prefimpl/pref.go:117]: Okay, upon further review, I think it's intentional in this particular method 😅

File: pkg/services/preference/prefimpl/pref.go [hunk 2]
```diff
@@ -141,6 +146,13 @@ func (s *Service) Patch(ctx context.Context, cmd *pref.PatchPreferenceCommand) e
 		exists = true
 	}
 
+	if cmd.Locale != nil {
+		if preference.JSONData == nil {
+			preference.JSONData = &pref.PreferenceJSONData{}
+		}
+		preference.JSONData.Locale = *cmd.Locale
```
Comment (idafurjes) [inline pkg/services/preference/prefimpl/pref.go:153]: is there a case when cmd.Locale is nil?

File: pkg/models/preferences.go [hunk 3]
```diff
@@ -87,6 +88,7 @@ type SavePreferencesCommand struct {
 	Timezone        string                  `json:"timezone,omitempty"`
 	WeekStart       string                  `json:"weekStart,omitempty"`
 	Theme           string                  `json:"theme,omitempty"`
+	Locale          string                  `json:"locale,omitempty"`
```
[the remark this answers is not in the document]
Comment (idafurjes) [reply pkg/models/preferences.go:91]: I don't think these structs are used anymore, maybe I should remove them

[the remark this answers is not in the document]
Comment (joshhunt) [reply pkg/models/preferences.go:91]: I think I did? https://github.com/grafana/grafana/blob/joshhunt/49424-local-user-pref/pkg/services/preference/model.go#L50

File: pkg/services/preference/prefimpl/pref.go (hunk 2 above)
Comment (sakjur) [reply pkg/services/preference/prefimpl/pref.go:153]: Not sure if that's new or was there when you commented, but the if right above this makes that check, so not on this line.

**Review by idafurjes — APPROVED** [review]
LGTM

**Comment by sakjur** [comment]
@joshhunt I've improved the unit tests a bunch now, fwiw.
If you want, I can add a unit test for you, or you can take a look and add it yourself.

**Comment by joshhunt** [comment]
Thanks @sakjur! I'll take a look at them!!!

**Pull request merged**